src/ai/contexts.cpp

```diff
@@ -1063,24 +1063,23 @@ double readonly_context_impl::power_projection(const map_location& loc, const mo
 
 			int64_t village_bonus = map_.is_village(terrain) ? 3 : 2;
 			int64_t defense = 100 - un.defense_modifier(terrain);
-			int64_t rating_64 = hp * defense * most_damage * village_bonus / 200;
-			int rating = rating_64;
-			if(static_cast<int64_t>(rating) != rating_64) {
+			int64_t rating = hp * defense * most_damage * village_bonus / 200;
+			if(rating > INT32_MAX) {
 				WRN_AI << "overflow in ai attack calculation";
 			}
 			if(rating > best_rating) {
 				map_location *pos = std::find(beg_used, end_used, it->second);
 				// Check if the spot is the same or better than an older one.
 				if (pos == end_used || rating >= ratings[pos - beg_used]) {
-					best_rating = rating;
+					best_rating = static_cast<int>(rating);
 					best_unit = it->second;
 				}
 			}
 		}
 
 		if (!best_unit.valid()) continue;
 		map_location *pos = std::find(beg_used, end_used, best_unit);
-		int index = pos - beg_used;
+		long index = pos - beg_used;
```

Comment:
Using `long` seems incorrect. If this is looking for a `signed size_t`, why not `std::ptrdiff_t`, the appropriate `different_type` or just `auto index = std::distance(...)`?

Reply:
This is precisely `difference_type`, which almost always expands to `ptrdiff_t`. Probably `auto` is good enough.

Reply:
Updated.